src/Raven.Server/Documents/CollectionRunner.cs

```diff
@@ -85,36 +85,42 @@ public virtual Task<IOperationResult> ExecuteDelete(string collectionName, long
                     {
                         foreach (var document in GetDocuments(context, collectionName, startEtag, internalQueryOperationStart, OperationBatchSize, isAllDocs, DocumentFields.Id))
                         {
-                            internalQueryOperationStart++;
+                            using (document)
+                            {
+                                internalQueryOperationStart++;
 
-                            cancellationToken.ThrowIfCancellationRequested();
+                                cancellationToken.ThrowIfCancellationRequested();
 
-                            token.Delay();
+                                token.Delay();
 
-                            if (isAllDocs && document.Id.StartsWith(HiLoHandler.RavenHiloIdPrefix, StringComparison.OrdinalIgnoreCase))
-                                continue;
+                                if (isAllDocs && document.Id.StartsWith(HiLoHandler.RavenHiloIdPrefix, StringComparison.OrdinalIgnoreCase))
+                                    continue;
 
-                            if (document.Etag > lastEtag) // we don't want to go over the documents that we have patched
-                            {
-                                end = true;
-                                break;
-                            }
+                                if (document.Etag > lastEtag) // we don't want to go over the documents that we have patched
+                                {
+                                    end = true;
+                                    break;
+                                }
 
-                            startEtag = document.Etag + 1;
+                                startEtag = document.Etag + 1;
 
-                            if (start > 0)
-                            {
-                                start--;
-                                continue;
-                            }
+                                if (start > 0)
+                                {
+                                    start--;
+                                    continue;
+                                }
 
-                            if (take-- <= 0)
-                            {
-                                end = true;
-                                break;
-                            }
+                                if (take-- <= 0)
+                                {
+                                    end = true;
+                                    break;
+                                }
 
-                            ids.Enqueue(document.Id);
+                                ids.Enqueue(document.Id);
+
+                                if (ids.Count >= OperationBatchSize)
+                                    break;
```

Comment:
The whole problem is that the collection query is:
1. Not supporting `DocumentField` - should be easy to do?
2. Shouldn't it support paging there, like the others inside 'GetDocuments' ?

Reply:
We already request `OperationBatchSize` in the `foreach`